service/matching/taskqueuemonitor.go

```diff
@@ -0,0 +1,116 @@
+// The MIT License
+//
+// Copyright (c) 2020 Temporal Technologies Inc.  All rights reserved.
+//
+// Copyright (c) 2020 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package matching
+
+import (
+	"context"
+	"math"
+	"time"
+)
+
+type (
+	tqmonStop    struct{}
+	tqmonGC      struct{}
+	tqmonUsurped struct{ id *taskQueueID }
+
+	// tqmonAccess is the interface by which a taskQueueMonitor observes and
+	// updates a set of taskQueueManager instances.
+	tqmonAccess interface {
+		getTaskQueues(int) []taskQueueManager
+		unloadTaskQueue(*taskQueueID)
+	}
+
+	// taskQueueMonitor performs garbage collection and unloads instances of
+	// taskQueueManager accessed via a tqmonAccess.
+	taskQueueMonitor struct {
+		sigs chan interface{}
+	}
+)
+
+// newTaskQueueMonitor constructs a new instance of taskQueueMonitor.
+func newTaskQueueMonitor() *taskQueueMonitor {
+	return &taskQueueMonitor{sigs: make(chan interface{})}
+}
+
+// Run runs the taskQueueMonitor event loop, performing GC of a set of
+// taskQueueManager instances and unloading the same when signaled or when GC
+// deems it appropriate. The implementaion of tqmonAccess MUST NOT re-enter this
+// taskQueueMonitor instance.
+func (m *taskQueueMonitor) Run(a tqmonAccess, c *Config) {
+	defer close(m.sigs)
+	t := time.NewTimer(c.IdleTaskqueueCheckInterval())
```

Comment:
NewTicker? or the idea here is to have the ability to get the updated (potentially) `IdleTaskqueueCheckInterval`?

Reply:
Yes, the re-fetch of the config is why I Reset() every time instead of using a Ticker.